engine/runtime/src/main/scala/org/enso/compiler/pass/resolve/FullyQualifiedNames.scala

```diff
@@ -0,0 +1,358 @@
+                  selfTypeResolution,
+                  pkgRepo
+                )
+              )
+          case _ =>
+            app.mapExpressions(
+              processExpression(
+                _,
+                bindings,
+                freshNameSupply,
+                selfTypeResolution,
+                pkgRepo
+              )
+            )
+
+        }
+
+    }
+
+  private def resolveLocalApplication(
+    app: IR.Application.Prefix,
+    bindings: BindingsMap,
+    freshNameSupply: FreshNameSupply,
+    pkgRepo: Option[PackageRepository],
+    selfTypeResolution: Option[Resolution]
+  ): IR.Expression = {
+    val processedFun =
+      processExpression(
+        app.function,
+        bindings,
+        freshNameSupply,
+        selfTypeResolution,
+        pkgRepo
+      )
+    val processedArgs =
+      app.arguments.map(
+        _.mapExpressions(
+          processExpression(
+            _,
+            bindings,
+            freshNameSupply,
+            selfTypeResolution,
+            pkgRepo
+          )
+        )
+      )
+
+    val processedApp = processedArgs match {
+      case List(thisArg) =>
+        (thisArg.value.getMetadata(this).map(_.target), processedFun) match {
+          case (Some(resolved @ ResolvedLibrary(_)), name: IR.Name.Literal) =>
+            resolveQualName(resolved, name, pkgRepo).fold(
+              err => Some(err),
+              _.map(resolvedMod =>
+                freshNameSupply
+                  .newName()
+                  .updateMetadata(this -->> resolvedMod)
+                  .setLocation(name.location)
+              )
+            )
+          case _ =>
+            None
+        }
+      case _ =>
+        None
+    }
+
+    processedApp.getOrElse(
+      app.copy(function = processedFun, arguments = processedArgs)
+    )
+  }
+
+  private def resolveQualName(
+    thisResolution: ResolvedLibrary,
+    consName: IR.Name.Literal,
+    optPkgRepo: Option[PackageRepository]
+  ): Either[IR.Expression, Option[FQNResolution]] = {
+    optPkgRepo
+      .flatMap { pkgRepo =>
+        val libName = LibraryName(thisResolution.namespace, consName.name)
+        pkgRepo.ensurePackageIsLoaded(libName).toOption.flatMap { _ =>
+          pkgRepo
+            .getLoadedModule(
+              s"${libName.toString}.${Imports.mainModuleName.name}"
+            )
+            .map { m =>
+              if (m.getIr == null) {
+                // Limitation of Fully Qualified Names:
+                // If the library has not been imported explicitly, then we won't have
+                // IR for it. Triggering a full compilation at this stage may have
+                // undesired consequences and is therefore prohibited on purpose.
+                Left(
+                  IR.Error.Resolution(
+                    consName,
+                    MissingLibraryImportInFQNError(thisResolution.namespace)
+                  )
+                )
+              } else {
```

Comment:
If we are not allowing FQNs of unimported modules, I don't think this is the right approach to implement it.

Please keep in mind that `ensurePackageIsLoaded` is potentially very costly operation - it may even _download_ a new library if the requested library is not yet cached.

Reply:
I appreciate that we want to distinguish 'module was not imported' vs 'module with such path does not exist in that library'.

But without importing (downloading) the library, we cannot really tell the second one reliably.

I would suggest then to add `isPackageLoaded` to the `PackageRepository` which just checks if it's loaded, and then do the following (in pseudocode):

```
if (pkgRepo.isPackageLoaded(libName)) {
   m = pkgRepo.getLoadedModule(s"${libName.toString}.${Imports.mainModuleName.name}")
   // do we still need to check if m.getIR == null? If it was imported it should already be compiled. I don't think at this point this scenario is viable but if it is, we can check
   Right(...)
} else {
   Left(MissingLibraryImport(libName))
}
```

That way we will avoid the costly package load/download.

Reply:
Fair comment. I will make the change